src/canonmn_int.cpp

```diff
@@ -442,11 +459,46 @@ namespace Exiv2 {
         TagInfo(0x00c1, "0x00c1", "0x00c1", N_("Unknown"), canonId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x00d0, "VRDOffset", N_("VRD Offset"), N_("VRD offset"), canonId, makerTags, unsignedLong, -1, printValue),
         TagInfo(0x00e0, "SensorInfo", N_("Sensor Info"), N_("Sensor info"), canonId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x2600, "AFInfoSize", N_("AF InfoSize"), N_("AF InfoSize"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x2601, "AFAreaMode", N_("AF Area Mode"), N_("AF Area Mode"), canonId, makerTags, signedShort, -1, EXV_PRINT_TAG(canonAFAreaMode)),
+        TagInfo(0x2602, "AFNumPoints", N_("AF NumPoints"), N_("AF NumPoints"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x2603, "AFValidPoints", N_("AF ValidPoints"), N_("AF ValidPoints"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x2604, "AFImageWidth", N_("AF ImageWidth"), N_("AF ImageWidth"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x2605, "AFImageHeight", N_("AF ImageHeight"), N_("AF ImageHeight"), canonId, makerTags, signedShort, -1, printValue),
```

Comment:
One last issue: `AFImageWidth` and `AFImageHeight` is a little confusing IMO, as it actually does not really have anything to do with AF (except for the fact that it is stored in AFInfo). Phil's documentation calls it `CanonImageWidth/Height` , however he called `AFWidth/Height` `AFImageWidth/Height`.

Calling it `AFWidth/Height` below is just fine. But I'd like to suggest renaming `AFImageWidth` to `ImageWidth` and `AFImageHeight` to `ImageHeight` .They are somewhat redundant, as they are always the same as `Exif.Image.ImageWidth/Height` and `Exif.Photo.PixelX/YDimension`. I am aware that this renaming would introduce a little inconsistency to the AF tags from Nikon, so I leave it up to you whether to omit the tags, rename them or leave them as they are.

Reply:
Very pleased with how this is working out.  If you'd like an unpaid job working on Exiv2, I'd be delighted to welcome you to the team!

I agree.  We should use the same names as ExifTool.  For sure the fields containing the actual image size are irrelevant and should not be there.  Let's call it CanonImageWidth/Height.

Thanks for the patch.  Nice code and works:
```
1385 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ bin/exiv2 -pa --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3
Exif.Canon.AFPointsInFocus                   SShort      6  20 21 25 45 46 47
Exif.Canon.AFPointsSelected                  SShort      9  19 20 21 24 25 26 45 46 47
Exif.Canon.AFPrimaryPoint                    SShort      0  
1386 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```
I'd like to ask you to change this a little.  There's a function in TagInfo to define how to print the value of a tag.  That's where we should be printing `20 21 25...`.
```
$ grep -Hn AFPointsInFocus src/canon*.cpp
../src/canonmn_int.cpp:474:        TagInfo(0x260c, "AFPointsInFocus", N_("AF Points in Focus"), N_("AF Points in Focus"), canonId, makerTags, signedShort, -1, printValue),
../src/canonmn_int.cpp:498:44	    AFPointsInFocus	    16
$ 
```

I think we're saying that Exif.Canon.AFPointsInFocus/AFPointsSelected are 4 unsignedShorts.  When you run exiv2 with the option *-pv* (v for value) `$exiv2 -pv --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3`, you see 4 unsigned shorts.  (the raw data)

When you run exiv2 with the option *-pt* (t for translated), your printing function will be called `$exiv2 -pt --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3` (the human interpreted data).

There's also something weirdo about Exif.Canon.AFPrimaryPoint.  It's claiming to be of length 0.  That can't be right.

Reply:
> If you'd like an unpaid job working on Exiv2, I'd be delighted to welcome you to the team!

Thanks, but I'm afraid I won't have time for this as I already have two maintainerships here at Github. Sry.

> I think we're saying that Exif.Canon.AFPointsInFocus/AFPointsSelected are 4 unsignedShorts.

In the raw exif data it's of variable length, depending how many AFPoints the camera has.

> When you run exiv2 with the option -pv (v for value) $exiv2 -pv --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3, you see 4 unsigned shorts. (the raw data)
> When you run exiv2 with the option -pt (t for translated), your printing function will be called $exiv2 -pt --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3 (the human interpreted data).

No matter how I run it, I always see the interpreted data.

> There's also something weirdo about Exif.Canon.AFPrimaryPoint. It's claiming to be of length 0. That can't be right.

I don't know what "PrimaryPoint" means. In my case it's always zero. Looking at the source code of exiftool, I can only assume that it's another bitmask. And when the bitmask is zero, there is no data to interpret, i.e. nothing to feed Exiv2::Value, and this results in a zero length of interpreted data. Not to be confused with the raw value.

Reply:
OK.  Leave this with me.  I'm happy with your code and I'll make the changes that I have in mind and update the PR by Monday.  

`TagInfo.printFct` prints the value.  For many tags it is `printValue()` which is the same for both **-pv** (value) and **-pt** (translated).  However, some tags such as AFAreaMode do change the presentation of the data:
```
$ exiv2 -pt --grep AreaMode ~/Downloads/AAAA5440.CR2 
Exif.Canon.AFAreaMode                        SShort      1  Zone AF
$ exiv2 -pv --grep AreaMode ~/Downloads/AAAA5440.CR2 
0x2601 Canon        AFAreaMode                  SShort      1  7
$ exiv2 -pv --grep AreaMode ~/Downloads/IMG_7600.jpg 
0x2601 Canon        AFAreaMode                  SShort      1  2
$ exiv2 -pt --grep AreaMode ~/Downloads/IMG_7600.jpg 
Exif.Canon.AFAreaMode                        SShort      1  Single-point AF
$
```
Here's the code:
```
1400 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ grep AFAreaMode src/canonmn_int.cpp 
    extern const TagDetails canonAFAreaMode[] = {
        TagInfo(0x2601, "AFAreaMode", N_("AF Area Mode"), N_("AF Area Mode"), canonId, makerTags, signedShort, -1, EXV_PRINT_TAG(canonAFAreaMode)),
$ 
```